### PR TITLE
docs: port manual source material to MyST markdown

### LINKED ISSUE #366: Port manual source material to MyST markdown
This would allow us to use the [MyST-NB](https://myst-nb.readthedocs.io/en/latest/) library to build html pages for both the manual and examples.

Currently the manual is built with [jupyter-sphinx](https://jupyter-sphinx.readthedocs.io/en/latest/) and the examples with MyST-NB which leads to some visual inconsistency in the code snippets between the two.

There are automated tools for doing the conversion

`rst-to-myst` -> https://github.com/executablebooks/rst-to-myst
`myst-parser` -> https://github.com/executablebooks/myst-parser

### PR DESCRIPTION
**Pull request by CalMacCQ**
# Description

Porting the manual source to MyST to allow easier use of the MyST library. We now use the MyST-nb library for building both the manual and examples pages. This ensures visual consistency of the example snippets.

Read the docs on markdown only notebooks -> https://myst-nb.readthedocs.io/en/latest/authoring/text-notebooks.html

Instructions for building the html locally -> https://github.com/CQCL/pytket-docs/tree/main/docs#building-the-docs (let me know if these are unclear)

I'd also recommend using the MyST VSCode extension in review -> https://marketplace.visualstudio.com/items?itemName=ExecutableBookProject.myst-highlight

## Preview

Screenshot 2024-08-29 at 17 00 23

Most of the conversion can be done in an automated way using [rst-to-myst](https://github.com/executablebooks/rst-to-myst)

```shell
rst2myst convert docs/manual/*.rst
```

This works for just about everything apart from `.. jupyter-execute::` directives which I rsubstitue with equivalent ` 
```
{code-cell} ipython3
```

sytax with VSCode find and replace.

## Example replacement for jupyter-execute directives
```
```{code-cell} ipython3

from pytket import Circuit
from pytket.extensions.qiskit import AerBackend
from pytket.utils import probs_from_counts

circ = Circuit(2, 2)
circ.H(0).X(1).measure_all()
backend = AerBackend()
compiled_circ = backend.get_compiled_circuit(circ)

handle = backend.process_circuit(compiled_circ, n_shots=2000)
counts = backend.get_result(handle).get_counts()
print(counts)

print(probs_from_counts(counts))
```

## Example replacement for jupyter-input/output directives

Input (truncated)
```
```{code-cell} ipython3
---
tags: [skip-execution]
---

from pytket import Circuit
from pytket.extensions.qiskit import IBMQBackend
from pytket.utils import expectation_from_counts

backend = IBMQBackend("ibmq_quito")

state = Circuit(3)
state.H(0).CX(0, 1).CX(1, 2).X(0)

```

Output

  ```
  (1.2047999999999999-0.0015000000000000013j)
  ```

## Example of a cell which raises an error
```
```{code-cell} ipython3
---
tags: [raises-exception]
---

from pytket import Circuit, Qubit, Bit

circ = Circuit()
# set up a circuit with qubit a[0]
circ.add_qubit(Qubit("a", 0))

# rejected because "a" is already a qubit register
circ.add_bit(Bit("a", 1))
```

# Related issues

closes #366

### CODE REVIEW

File: docs/manual/manual_backend.md [hunk 1]
```diff
@@ -0,0 +1,850 @@
...
+print(dev_b.required_predicates)
+print(sim_b.required_predicates)
+```
+
+
+```
+
+    [NoClassicalControlPredicate, NoFastFeedforwardPredicate, NoMidMeasurePredicate, NoSymbolsPredicate, GateSetPredicate:{ U1 noop U2 CX Barrier Measure U3 }, DirectednessPredicate:{ Nodes: 5, Edges: 8 }]
+    [NoClassicalControlPredicate, NoFastFeedforwardPredicate, GateSetPredicate:{ CU1 CZ CX Unitary2qBox Sdg U1 Unitary1qBox SWAP S U2 CCX Y U3 Z X T noop Tdg Reset H }]
+```
```
Comment (willsimmons1465) [inline docs/manual/manual_backend.md:55]: Remove the extra line and tabs

File: docs/manual/manual_backend.md [hunk 2]
```diff
@@ -0,0 +1,850 @@
...
+counts = backend.get_result(handle).get_counts()
+print(counts)
+```
+
+
+```
+
+{(0, 1, 1): 1000}
+
+```
```
Comment (willsimmons1465) [inline docs/manual/manual_backend.md:817]: Remove the extra lines

File: docs/manual/manual_compiler.md [hunk 3]
```diff
@@ -0,0 +1,1265 @@
...
+print(r0.get_counts())
+print(r1.get_counts())
+```
+
+
+```
+
+{(0, 0, 0, 0): 503, (0, 0, 0, 1): 488, (0, 1, 0, 0): 533, (0, 1, 0, 1): 493, (1, 0, 0, 0): 1041, (1, 0, 0, 1): 107, (1, 0, 1, 0): 115, (1, 0, 1, 1): 14, (1, 1, 0, 0): 576, (1, 1, 0, 1): 69, (1, 1, 1, 0): 54, (1, 1, 1, 1): 7}
+{(0, 0, 0, 0): 2047, (0, 1, 0, 0): 169, (0, 1, 1, 0): 1729, (1, 1, 0, 0): 7, (1, 1, 1, 0): 48}
+```
```
Comment (willsimmons1465) [inline docs/manual/manual_compiler.md:1007]: Blank line here

File: docs/manual/manual_noise.md [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+
+print(backend.backend_info.all_node_gate_errors[Node(0)])
+print(backend.backend_info.all_node_gate_errors[Node(1)])
+```
+
+
+```
+
+{<OpType.noop: 55>: 0.00036435993708370417,
+<OpType.Rz: 32>: 0.0,
```
Comment (willsimmons1465) [inline docs/manual/manual_noise.md:107]: Blank line

File: docs/manual/manual_noise.md [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+{(node[4], node[3]): {<OpType.CX: 37>: 0.01175674116384029},
+(node[3], node[4]): {<OpType.CX: 37>: 0.005878370581920145},
+(node[2], node[3]): {<OpType.CX: 37>: 0.013302220876095505},
+(node[3], node[2]): {<OpType.CX: 37>: 0.006651110438047753},
+(node[2], node[1]): {<OpType.CX: 37>: 0.022572084465386333},
+(node[1], node[2]): {<OpType.CX: 37>: 0.011286042232693166},
+(node[0], node[1]): {<OpType.CX: 37>: 0.026409836177538337},
+(node[1], node[0]): {<OpType.CX: 37>: 0.013204918088769169}}
+
+```
```
Comment (willsimmons1465) [inline docs/manual/manual_noise.md:136]: Blank lines

File: docs/manual/manual_noise.md [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+print('\nGraphPlacement mapping:')
+for k, v in graph_placement.items():
+    print(k, v)
+
+```
+
+
+```
+
+[(node[0], node[1]), (node[1], node[0]), (node[1], node[2]), (node[1], node[3]), (node[2], node[1]), (node[3], node[1]), (node[3], node[4]), (node[4], node[3])]
```
Comment (willsimmons1465) [inline docs/manual/manual_noise.md:191]: Blank line

File: docs/manual/manual_noise.md [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+```
+
+
+```
+
+Noiseless Counts: Counter({(1, 1): 6415, (1, 0): 6385})
+Base Noisy Counts: Counter({(1, 0): 6368, (1, 1): 5951, (0, 1): 253, (0, 0): 228})
+Recombined Noisy Counts using PauliFrameRandomisation: {(0, 1): 203, (0, 0): 215, (1, 0): 6194, (1, 1): 6188}
+
+```
```
Comment (willsimmons1465) [inline docs/manual/manual_noise.md:344]: Blank lines

File: docs/manual/manual_noise.md [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+
+
+```
+
+Noiseless Counts: Counter({(1, 0): 6490, (1, 1): 6310})
+Recombined Noiseless Counts using UniversalFrameRandomisation: {(1, 0): 6440, (1, 1): 6360}
+Base Noisy Counts: Counter({(1, 0): 6298, (1, 1): 6022, (0, 1): 261, (0, 0): 219})
+Recombined Noisy Counts using PauliFrameRandomisation: {(0, 1): 240, (0, 0): 212, (1, 0): 6253, (1, 1): 6095}
+Recombined Noisy Counts using UniversalFrameRandomisation: {(0, 1): 208, (0, 0): 208, (1, 0): 6277, (1, 1): 6107}
+```
```
Comment (willsimmons1465) [inline docs/manual/manual_noise.md:398]: Blank lines

File: docs/manual/manual_noise.md [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+    averaging_circuits = frame_randomisation.get_all_circuits(circ)
+    print('For a single gate in the averaging ensemble we return a single circuit:')
+    for com in averaging_circuits[0]:
+        print(com)
+
+    print('\nWe can check that the unitary of the circuit is preserved by comparing output counts:')
+    backend = AerBackend()
+    print(backend.run_circuit(circ, 100).get_counts())
+    print(backend.run_circuit(averaging_circuits[0], 100).get_counts())
+```
```
Comment (willsimmons1465) [inline docs/manual/manual_noise.md:256]: Indenting here

File: docs/manual/manual_noise.md [hunk 10]
```diff
@@ -0,0 +1,543 @@
...
+    ]
+    # combine each averaging circuits counts into a single counts object for comparison
+    pfr_counts = {}
+    for counts in pfr_counts_list:
+        pfr_counts = {key: pfr_counts.get(key,0) + counts.get(key,0) for key in set(pfr_counts)|set(counts)}
+
+    print(pfr_counts)
+    print(backend.run_circuit(circ, 50*len(averaging_circuits)).get_counts())
+
+```
```
Comment (willsimmons1465) [inline docs/manual/manual_noise.md:301]: Indenting here

File: docs/manual/manual_noise.md [hunk 11]
```diff
@@ -0,0 +1,543 @@
...
+    noiseless_result = noiseless_backend.get_result(noiseless_handle)
+
+    circ_parallel_measure = spam_correcter.get_parallel_measure(circ)
+    corrected_counts = spam_correcter.correct_counts(noisy_result, circ_parallel_measure)
+
+    print('Noisy Counts:', noisy_result.get_counts())
+    print('Corrected Counts:', corrected_counts.get_counts())
+    print('Noiseless Counts:', noiseless_result.get_counts())
+
+```
```
Comment (willsimmons1465) [inline docs/manual/manual_noise.md:537]: Indenting here

File: docs/manual/manual_compiler.md [hunk 12]
```diff
@@ -0,0 +1,1265 @@
...
+    midmeasure = NoMidMeasurePredicate()
+
+    print(gateset.verify(circ))
+    print(midmeasure.verify(circ))
+
+    circ.S(0)
+
+    print(gateset.verify(circ))
+    print(midmeasure.verify(circ))
+```
```
Comment (willsimmons1465) [inline docs/manual/manual_compiler.md:48]: Indenting

**Review by willsimmons1465 — CHANGES_REQUESTED** [review]
Nicely done! Some of the code cells are inconsistently formatted with whitespace, so the only changes needed are going through to fix that.

I'm presuming from the existing content using MyST that it fully works with CI, i.e. if a cell fails to run CI would report an error?

File: docs/manual/manual_backend.md (hunk 1 above)
Comment (CalMacCQ) [reply docs/manual/manual_backend.md]: Thanks for spotting these formatting issues.

I fixed them in [ad08277](https://github.com/CQCL/pytket-docs/pull/370/commits/ad08277e4175c58db1b1eb5409f49998540553e7)

**Comment by CalMacCQ** [comment]
Added some code which errored out in [0bb46c3](https://github.com/CQCL/pytket-docs/pull/370/commits/0bb46c3e9b6ebf34d7eda68b6a9ed5e188724fa1) and this caused C.I. to fail with a `CellExecutionError` which is reassuring.

File: docs/manual/manual_circuit.md [hunk 13]
```diff
@@ -0,0 +1,1935 @@
...
+
+Its also possible to count all the occurrences of each {py:class}`~pytket.circuit.OpType` using the {py:func}`~pytket.utils.stats.gate_counts` function from the {py:mod}`pytket.utils` module.
+
+
+```{code-cell} ipython3
+
+    from pytket.utils.stats import gate_counts
+
+    gate_counts(circ)
+```
```
Comment (willsimmons1465) [inline docs/manual/manual_circuit.md:669]: I missed this one in the last review

**Review by willsimmons1465 — CHANGES_REQUESTED** [review]

File: docs/manual/manual_circuit.md (hunk 13 above)
Comment (CalMacCQ) [reply docs/manual/manual_circuit.md]: Thanks Will, fixed in [4b7448b](https://github.com/CQCL/pytket-docs/pull/370/commits/4b7448be05e9078e563f542668fafb25ca241328)

Comment (CalMacCQ) [reply docs/manual/manual_circuit.md]: I did a search to try and rule out other indentation issues, didn't find any biut maybe the search pattern could be wrong.

**Review by willsimmons1465 — APPROVED** [review]
Looks good to go! Thanks for sorting this out

**Pull request merged**